### PR TITLE
Modularize API calls

### PR DESCRIPTION
**Pull request by jdesboeufs**

### CODE REVIEW

File: src/CatalogDetail/CatalogDetail.js [hunk 1]
```diff
@@ -5,49 +5,55 @@ import CatalogSection from '../Section/CatalogSection/CatalogSection'
...
+
+  updateMetrics() {
+    return fetchMetrics(this.props.params.catalogId)
+      .then(metrics => {
+        this.setState({ metrics })
       })
-      .catch((err) => {
-        console.error(err)
+      .catch(err => {
+        if (this.state.errors.indexOf(err.message) < 0) {
```
Comment (jdesboeufs) [inline src/CatalogDetail/CatalogDetail.js]: Use `Array.includes`.

File: src/CatalogDetail/CatalogDetail.js [hunk 2]
```diff
@@ -5,49 +5,55 @@ import CatalogSection from '../Section/CatalogSection/CatalogSection'
...
-          console.error(err)
-        })
-      }
+  updateCatalog() {
+    return fetchCatalog(this.props.params.catalogId)
+      .then(catalog => {
+        this.setState({ catalog })
+      })
+      .catch(err => {
+        if (this.state.errors.indexOf(err.message) < 0) {
```
Comment (jdesboeufs) [inline src/CatalogDetail/CatalogDetail.js]: Same.

File: src/CatalogDetail/CatalogDetail.js [hunk 3]
```diff
@@ -5,49 +5,55 @@ import CatalogSection from '../Section/CatalogSection/CatalogSection'
...
-        .catch((err) => {
-          console.error(err)
-        })
-      }
+  updateCatalog() {
+    return fetchCatalog(this.props.params.catalogId)
+      .then(catalog => {
+        this.setState({ catalog })
+      })
+      .catch(err => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/CatalogDetail.js:41]: Can be extracted as a helper

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+describe('<CatalogDetail />', () => {
+
+  describe('When all goes well', () => {
+
+    it('should render CatalogDetail', () => {
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+
+      wrapper.setState({catalog, metrics})
+      expect(wrapper.find('#catalog-detail').length).toEqual(1)
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js:17]: Cheap

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        .updateCatalog()
+        .then(() => expect(wrapper.state('catalog')).toEqual(catalog))
+    })
+
+    it('should set catalog and metrics', () => {
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+      return wrapper
+        .instance()
+        .componentWillMount()
+        .then(() => expect(wrapper.state('metrics')).toEqual(metrics))
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: Missing test for catalog part

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+
+      wrapper.setState({catalog, metrics})
+      expect(wrapper.find('#catalog-detail').length).toEqual(1)
+    })
+  })
+
+  describe('fetch', () => {
+
+    it('should set metrics', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: No description

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+
+      return wrapper
+        .instance()
+        .updateMetrics()
+        .then(() => expect(wrapper.state('metrics')).toEqual(metrics))
+    })
+
+    it('should set catalog', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: No description

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+      return wrapper
+        .instance()
+        .updateCatalog()
+        .then(() => expect(wrapper.state('catalog')).toEqual(catalog))
+    })
+
+    it('should set catalog and metrics', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: No description

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '1'}} />)
+      return wrapper
+        .instance()
+        .componentWillMount()
+        .then(() => expect(wrapper.state('metrics')).toEqual(metrics))
+    })
+
+    it('should render an empty div', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: Same

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '0'}} />)
+      expect(wrapper.html()).toEqual("<div></div>")
+    })
+  })
+
+  describe('Error', () => {
+    describe('Metrics error', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: Poor description

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+          .then(() => expect(wrapper.state('errors')).toContain("catalogId is required"))
+      })
+
+      it('Should return an error, metrics not found', () => {
+        const wrapper = shallow(<CatalogDetail params={{catalogId: '0'}} />)
+
+        return wrapper
+          .instance()
+          .updateMetrics()
+          .then(() => expect(wrapper.state('errors')).toContain("Metrics not found"))
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js:48]: Not found is a technical error. You must explain to the final user

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+      })
+    })
+
+    describe('Catalog error', () => {
+      it('Should return an error, catalogId is required', () => {
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js]: Internal error. Should not be displayed but reported.
Many ways to deal with this.

File: src/CatalogDetail/CatalogDetail.js [hunk 13]
```diff
@@ -5,49 +5,55 @@ import CatalogSection from '../Section/CatalogSection/CatalogSection'
...
+        if (this.state.errors.indexOf(err.message) < 0) {
+          const errors = [...this.state.errors, err.message]
+          this.setState({ errors })
+        }
+      })
   }
 
   render() {
-    if (this.state.catalog && this.state.metrics) {
+    if (this.state && this.state.catalog && this.state.metrics) {
```
Comment (jdesboeufs) [inline src/CatalogDetail/CatalogDetail.js]: Test on `this.state` is now useless. You already use `this.state.something` elsewhere.

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    it('should render an empty div', () => {
+      const wrapper = shallow(<CatalogDetail params={{catalogId: '0'}} />)
+      expect(wrapper.html()).toEqual("<div></div>")
+    })
+  })
+
+  describe('Error', () => {
+    describe('Metrics error', () => {
+      it('Should return an error, catalogId is required', () => {
+        const wrapper = shallow(<CatalogDetail params={{}}/>)
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js:34]: Can we avoid `params={{}}`?

File: src/CatalogDetail/__test__/CatalogDetail.test.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    })
+
+    describe('Catalog error', () => {
+      it('Should return an error, catalogId is required', () => {
+        const wrapper = shallow(<CatalogDetail params={{}}/>)
+
+        return wrapper
+          .instance()
+          .updateMetrics()
+          .then(() => expect(wrapper.state('errors')).toContain("catalogId is required"))
```
Comment (jdesboeufs) [inline src/CatalogDetail/__test__/CatalogDetail.test.js:67]: Mixed single and double quotes. Can you enforce the ESLint conf?

File: src/fetch/fetch.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+function _f(url) {
+  return fetch(url)
+    .then(response => response.json())
+    .catch((err) => {
+      console.error(err)
```
Comment (jdesboeufs) [inline src/fetch/fetch.js:5]: `console.error`?
Finally this helper is always resolved, sometimes with `undefined`.
Is it the expected behavior?

File: src/CatalogDetail/__test__/CatalogDetail.test.js (hunk 15 above)
Comment (tmerlier) [reply src/CatalogDetail/__test__/CatalogDetail.test.js:67]: Ready! PR #42 `Enforce eslint quotes rules`

**Pull request merged**